File: src/leetcode/stack/RemoveDuplicates.java
package leetcode.stack;

import java.util.Stack;

/**
 * @author chengzw
 * @description 删除字符串中的所有相邻重复项，https://leetcode-cn.com/problems/remove-all-adjacent-duplicates-in-string/
 * <p>
 * 给出由小写字母组成的字符串 S，重复项删除操作会选择两个相邻且相同的字母，并删除它们。
 * 在 S 上反复执行重复项删除操作，直到无法继续删除。
 * <p>
 * 输入："abbaca"
 * 输出："ca"
 * 解释：
 * 例如，在 "abbaca" 中，我们可以删除 "bb" 由于两字母相邻且相同，这是此时唯一可以执行删除操作的重复项。
 * 之后我们得到字符串 "aaca"，其中又只有 "aa" 可以执行重复项删除操作，所以最后的字符串为 "ca"。
 * @since 2021/9/6
 */
public class RemoveDuplicates {

    public static String solution1(String s) {
        Stack<Character> stack = new Stack<>();
        int n = s.length();
        int i = 0;
        while (i < n) {
            char c = s.charAt(i);
            if (stack.isEmpty() || stack.peek() != c) {
                stack.push(c);
            } else {
                while (!stack.isEmpty() && stack.peek() == c) {
                    stack.pop();
                }
            }
            i++;
        }
        StringBuilder sb = new StringBuilder();
        while (!stack.isEmpty()) {
            sb.append(stack.pop());
        }
        return sb.reverse().toString();
    }

    public static void main(String[] args) {
        System.out.println(solution1("abbaca"));
    }
}

File: go/leetcode/stack/remove_duplicates.go
package stack

/**
 * @description 1047. 删除字符串中的所有相邻重复项
 * @author chengzw
 * @since 2022/6/1
 * @link https://leetcode.cn/problems/remove-all-adjacent-duplicates-in-string/
 */

/**
给出由小写字母组成的字符串 S，重复项删除操作会选择两个相邻且相同的字母，并删除它们。
在 S 上反复执行重复项删除操作，直到无法继续删除。
在完成所有重复项删除操作后返回最终的字符串。答案保证唯一。

示例：
输入："abbaca"
输出："ca"

解释：
例如，在 "abbaca" 中，我们可以删除 "bb" 由于两字母相邻且相同，这是此时唯一可以执行删除操作的重复项。
之后我们得到字符串 "aaca"，其中又只有 "aa" 可以执行重复项删除操作，所以最后的字符串为 "ca"。
*/

/**
思路：
1.一旦新来的字符和栈顶的字符一样的话，就弹出栈顶字符，直至扫完整个字符串。栈中剩下的字符串就是最终要输出的结果。

时间复杂度：O(n)
空间复杂度：O(1)
*/
func removeDuplicates(s string) string {
	stack := []rune{}
	for _, c := range s {
		if len(stack) == 0 || stack[len(stack)-1] != c {
			stack = append(stack, c)
		} else {
			stack = stack[:len(stack)-1]
		}
	}
	return string(stack)
}
